R/versioning.R

```diff
@@ -0,0 +1,90 @@
+#' Turn bucket versioning on or off, check status (default), or
+#' list archived versions of objects in the bucket and view their generation numbers.
+#'
+#'
+#' @param action "status", "enable", "disable", or "list"
+#' @param bucket gcs bucket
+#'
+#' @return versioned_objects dataframe #only if "list" action
+#' @importFrom googleAuthR gar_api_generator
+#' @family managementAPI functions
+#' @export
+gcs_version_bucket <- function(bucket, action = c("status", "enable", "disable", "list")) {
+  action <- match.arg(action)
+
+  bucket <- as.bucket_name(bucket)
+
+  # check that it is a valid bucket that user has access to
+  tryCatch({
+    gcs_get_bucket(bucket)
+    # cat(sprintf("\n \"%s\" is a valid bucket.\n", bucket))
+  }, error = function(e) {
+    stop("Bucket not found. Check bucket name and if you have read permissions.
+    Looked for ", bucket, call. = FALSE)
+  })
+
+  # will only be different for list
+  url <- sprintf(
+    "https://www.googleapis.com/storage/v1/b/%s",
+    bucket
+  )
+  pars_args <- list(fields = "versioning")
+
+  # Default is to check status
+  if (action == "status") {
+    api <- googleAuthR::gar_api_generator(url,
+      "GET",
+      pars_args = pars_args,
+      data_parse_function = function(x) x,
+      checkTrailingSlash = FALSE
+    )
+
+    if (length(api()) == 0) { # If versioning has never been enabled before
+      cat(sprintf("Versioning is NOT ENABLED for \"%s\"", bucket))
+    } else {
+      if (api()$versioning$enabled == TRUE) {
+        cat(sprintf("Versioning is ENABLED for \"%s\"", bucket))
+      } else {
+        cat(sprintf("Versioning is NOT ENABLED for \"%s\"", bucket))
+      }
+    }
+  }
+  # on or off
+  else if (action %in% c("enable", "disable")) {
+    trueorfalse <- ifelse(action == "enable", TRUE, FALSE)
+
+    body <- list(versioning = list(enabled = trueorfalse))
+
+    api <- googleAuthR::gar_api_generator(url,
+      "PATCH",
+      pars_args = pars_args,
+      data_parse_function = function(x) x,
+      checkTrailingSlash = FALSE
+    )
+
+    api(the_body = body)
+  }
+  else if (action == "list") {
+    # To list both live and archived versions of an object and view their generation numbers 
+    # (filter only archived versions for final output)
+    # Archived versions of objects have a `timeDeleted` property.
+
+    url <- sprintf(
+      "https://www.googleapis.com/storage/v1/b/%s/o",
+      bucket
+    )
+    pars_args <- list(versions = "true")
+
+    api <- googleAuthR::gar_api_generator(url,
+      "GET",
+      pars_args = pars_args,
+      data_parse_function = function(x) x$items,
+      checkTrailingSlash = FALSE
+    )
+    if (!("timeDeleted" %in% colnames(api()))) {
```

Comment:
This logic should ideally be in the `data_parse_function` - create a new utility function if its too big for here. 

I don't think it should raise an error if no archived objects are present otherwise users will have to wrap it in `try()` within scripts.

I took a look and I think the `list` functionality is the same as `gcs_list_objects()` ( https://github.com/cloudyr/googleCloudStorageR/blob/master/R/objects.R#L30 ) so it should call that function instead if needed - you are missing logic such as paging through all the objects for big buckets which that handles.

Reply:
So you could just remove the `list` parameter and code blocks, and any changes required can be done in `gcs_list_objects()` instead.

Reply:
Makes sense, no need to reinvent the wheel. Will do.

Reply:
Currently, `gcs_list_objects()` doesn't return **timeDeleted** column. Should that function be updated to return 19 columns instead of 18? I believe this is the correct part of the function parse_lo() helper function that would need to be updated:

```
  ## extra columns for composite objects (#73)
  x$componentCount  <- if(is.null(x$componentCount)) NA else x$componentCount
  x$contentLanguage <- if(is.null(x$contentLanguage)) NA else x$contentLanguage
  
 ## extra column for deleted objects (if bucket versioning is enabled)  (#96)
  x$timeDeleted <- if(is.null(x$timeDeleted)) NA else x$timeDeleted

  attr(x, "nextPageToken") <- nextPageToken
  attr(x, "prefixes") <- prefixes
```
If you agree with this strategy, then I don't see any value to even have the list functionality in this versioning function. Maybe status, enable, and disable actions is sufficient?

Reply:
Yes I agree, stick to status, enable and disable

Reply:
So, can you merge this pull? Then I can test out updating the existing list function in a separate pull or does it make sense to update that function first? Does any other function rely on the current schema of 18 columns? I feel like it could be a breaking change to add one more column.